fsspec/implementations/prefix.py

```diff
@@ -0,0 +1,196 @@
+from typing import Any, Iterable, Sequence, Union
+
+import fsspec
+from fsspec import AbstractFileSystem
+from fsspec.core import split_protocol
+from fsspec.spec import AbstractBufferedFile
+from fsspec.utils import stringify_path
+
+
+class PrefixBufferedFile(AbstractBufferedFile):
+    def _fetch_range(self, start, end):
+        pass
+
+
+class PrefixFileSystem(AbstractFileSystem):
+    """A meta-filesystem to add a prefix and delegate to another filesystem"""
+
+    def __init__(
+        self,
+        prefix: str,
+        fs: fsspec.AbstractFileSystem,
+        *args,
+        **storage_options,
+    ) -> None:
+        """
+        Parameters
+        ----------
+        prefix: str
+            The prefix to append to all paths
+
+        fs: fsspec.AbstractFileSystem
+            An instantiated filesystem to wrap. All operations are delegated to
+            this filesystem after appending the specified prefix
+        """
+        super().__init__(*args, **storage_options)
+        self.prefix = stringify_path(prefix)
+
+        if not self.prefix:
+            self.prefix = self.root_marker
+
+        self.fs = fs
+
+    def _get_relative_path(self, path: str) -> str:
+        if path[: len(self.sep)] == self.sep:
+            return path[len(self.sep) :]
+        return path
+
+    def _add_fs_prefix(self, path: str) -> Union[str, Sequence[str]]:
+        if isinstance(path, str):
+            path = stringify_path(path)
+            protocol, path = split_protocol(path)
+
+            path = self._get_relative_path(path)
+
+            if self.prefix == self.sep:
```

Comment:
I guess this is again used in `root` sense, right?

Reply:
Mmmh yes, but the distinction between `root_marker` and `sep` is a bit confusing

So I guess I should use `root_marker` also here?

https://github.com/intake/filesystem_spec/blob/2d3179b401b019f6d9cb28be018a34f6e9a7e28c/fsspec/implementations/prefix.py#L43-L46

Reply:
Following this conversation, I do tend to go back to my original argument, that a prefixfilesystem with no prefix should be an exception, and then you don't have to deal with this ambiguity.

Reply:
Ok I (think) I fixed several issues with the prefix/sep/root_marker

- If there is not prefix I raise a `ValueError`
- Any trailing `sep` on prefix is removed
- If prefix is equal to the `root_marker` add only the root marker once
- Use the `sep` and `root_marker` of the wrapped fs